refactor: Modernize build, enhance demo app, and prepare for archival

dkhawk opened the pull request:
This pull request introduces a series of significant updates to modernize the project, improve the demo application, and streamline the library structure in preparation for archiving.

### Key Changes:

*   **Build Modernization:**
    *   Upgraded Gradle to 8.13.2 and Kotlin to 2.3.0.
    *   Standardized `compileSdk` (36) and `minSdk` across all modules using the version catalog.
    *   Resolved various build deprecation warnings.
    *   Added a basic testing infrastructure with dependencies for JUnit, Mockito, and Mockk.

*   **Demo App UI/UX Enhancements:**
    *   Implemented a sticky immersive, edge-to-edge display for a more modern user experience.
    *   Added UI components to demonstrate "Find Current Place" and "Fetch Photo" functionality from the Places SDK.

*   **Breaking Change: Internalized Shared Code:**
    *   The `shared` module has been removed. Its utility classes (like `MainThreadObservable`) have been copied directly into the `maps-rx` and `places-rx` modules as `internal` classes.
    *   **Reasoning:** This eliminates the need for a separate Maven artifact for the shared code, simplifying the dependency graph for consumers who might install the AARs manually. This structural change is a prerequisite for archiving the repository.

This prepares the repository for its final, archived state by ensuring the libraries are self-contained and the demo app is a robust example of their usage.

dkhawk committed:
refactor: standardize build configuration and SDK versions

- Extract `compileSdk`, `minSdk`, `targetSdk` to version catalog and set `compileSdk`/`targetSdk` to 36.
- Fix build deprecations: replace `buildDir` with `layout.buildDirectory` and move `kotlinOptions` to `compilerOptions`.
- Update Kotlin to 2.3.0 and Gradle to 8.13.2.
- Switch app module to use local project dependencies (`:maps-rx`, `:places-rx`) instead of version catalog artifacts.
- Remove unused `ExampleInstrumentedTest`.
- Update .gitignore to exclude secrets.properties.

dkhawk committed:
feat: enhance demo app and add testing infrastructure

- Fix lint/deprecation warnings in MainActivity and PlacesRx
- Expand demo app with FindCurrentPlace and FetchPhoto UI
- Update build files to JDK 17
- Add testing dependencies (JUnit, Mockito, Mockk)
- Add initial (ignored) unit tests for maps-rx and places-rx

dkhawk committed:
fix(ui): enforce edge-to-edge with NoActionBar theme

dkhawk committed:
feat(ui): implement sticky immersive mode using WindowInsetsControllerCompat

dkhawk committed:
refactor: use clean imports in MainActivity

dkhawk committed:
refactor!: internalize shared code and prepare for archive

Removes the 'shared' module and duplicates its utility classes into 'maps-rx' and 'places-rx' as internal extensions.

This change eliminates the need for a separate 'shared' Maven artifact, allowing 'maps-rx' and 'places-rx' to be used as standalone libraries without complex transitive dependencies for manual AAR consumers. Code duplication was chosen over a shared artifact to simplify distribution as we prepare to archive these libraries.

Key changes:
- Deleted 'shared' module and updated settings.gradle.
- Copied MainThreadObservable and related utilities to 'com.google.maps.android.rx.maps.internal' and 'com.google.maps.android.rx.places.internal'.
- Updated all imports to use the local internal versions.
- Removed project dependency on ':shared' from build files.
- Marked all duplicated code as 'internal' to prevent binary compatibility issues or duplicate class definitions if both libraries are used together.

This refactoring streamlines the project structure in anticipation of archiving the repository.

kikoso commented on app/build.gradle.kts:
```diff
@@ -49,36 +49,39 @@ android {
-    }
-
-    kotlinOptions {
-        jvmTarget = "1.8"
```
Should we keep the previous Java version, to avoid any issues?

kikoso commented:
We would need to mark this PR as fix to trigger a new build.

kikoso committed:
chore: removed lint from old module

kikoso commented on places-rx/src/main/java/com/google/maps/android/rx/places/internal/MainThreadCompletable.kt:
```diff
@@ -0,0 +1,39 @@
+// Copyright 2021 Google LLC
```
What about renaming this module to places/internal? We need a different name to avoid conflicts with the maps-rx module.

kikoso commented on places-rx/build.gradle.kts:
```diff
@@ -16,32 +16,36 @@
     compileOptions {
-        sourceCompatibility = JavaVersion.VERSION_1_8
-        targetCompatibility = JavaVersion.VERSION_1_8
+        sourceCompatibility = JavaVersion.VERSION_17
```
What about keeping the previous Java version?

kikoso commented on places-rx/build.gradle.kts:
```diff
@@ -16,32 +16,36 @@
-        freeCompilerArgs += "-Xexplicit-api=strict"
+tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().configureEach {
+    compilerOptions {
+        jvmTarget.set(org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17)
```
Previous Java version.

kikoso reviewed (approved):
Some minor comments, most specifically about keeping the previous Java version. This seems to work if packed as an aar.

dkhawk committed:
Revert Java version bump back to Java 8

dkhawk commented:
Addressed the review comments:
- Reverted `sourceCompatibility`, `targetCompatibility`, and Kotlin `jvmTarget` back to Java `1.8` in `app`, `maps-rx`, and `places-rx` to maintain stability.
- Verified that `MainThreadCompletable` in `places-rx` is correctly isolated inside `com.google.maps.android.rx.places.internal`, meaning no collision exists with `maps-rx`.
Tests and connected tests confirmed passing.
